src/spikeinterface/sortingcomponents/peak_detection.py

```diff
@@ -105,6 +109,224 @@ def detect_peaks(recording, method='by_channel', pipeline_nodes=None,
+        PeakDetector.__init__(self, recording, return_output=return_output)
+        self.peak_detector_node = peak_detector_node
+        self.waveform_extraction_node = waveform_extraction_node
+        self.waveform_denoising_node = waveform_denoising_node
+        self.num_iterations = num_iterations
+
+    def get_trace_margin(self) -> int:
+        """
+        Calculate the maximum trace margin from the internal pipeline.
+        Using the strategy as use by the Node pipeline
+
+
+        Returns
+        -------
+        int
+            The maximum trace margin.
+        """
+        internal_pipeline = (self.peak_detector_node, self.waveform_extraction_node, self.waveform_denoising_node)
+        pipeline_margin = (node.get_trace_margin() for node in internal_pipeline if hasattr(node, "get_trace_margin"))
+        return max(pipeline_margin)
+
+    def compute(self, traces_chunk, start_frame, end_frame, segment_index, max_margin) -> Tuple[np.ndarray, np.ndarray]:
+        """
+        Perform the iterative peak detection, waveform extraction, and denoising.
+
+        Parameters
+        ----------
+        traces_chunk : array-like
+            The chunk of traces to process.
+        start_frame : int
+            The starting frame for the chunk.
+        end_frame : int
+            The ending frame for the chunk.
+        segment_index : int
+            The segment index.
+        max_margin : int
+            The maximum margin for the traces.
+
+        Returns
+        -------
+        tuple of ndarray
+            A tuple containing a single ndarray with the detected peaks.
+        """
+
+        traces_chunk = traces_chunk.astype("float32")
+        local_peaks_list = []
+        all_waveforms = []
+
+        for iteration in range(self.num_iterations):
+            (local_peaks,) = self.peak_detector_node.compute(
+                traces=traces_chunk, 
+                start_frame=start_frame, 
+                end_frame=end_frame,
+                segment_index=segment_index,
+                max_margin=max_margin,
+            )
+
+            local_peaks = self.add_iteration_to_peaks_dtype(local_peaks=local_peaks, iteration=iteration)
+            local_peaks_list.append(local_peaks)
+
+            # End algorith if no peak is found
+            if local_peaks.size == 0:
+                break
+
+            waveforms = self.waveform_extraction_node.compute(traces=traces_chunk, peaks=local_peaks)
+            denoised_waveforms = self.waveform_denoising_node.compute(
+                traces=traces_chunk, 
+                peaks=local_peaks, 
+                waveforms=waveforms
+            )
+            traces_chunk_minus_waveforms, _ = self.substract_waveforms_from_traces(
+                traces_chunk, local_peaks, 
+                denoised_waveforms
+            )
+
+            # update traces_chunk to run the algorithm again and store local_peaks
+            traces_chunk = traces_chunk_minus_waveforms
```

Comment:
Here we could do a inplace substraction instead of creating a new array and changing the ref at each iteration. No ?